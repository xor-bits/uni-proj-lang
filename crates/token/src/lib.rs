#[cfg(test)]
use serde::{ser::SerializeStruct, Serialize};
use std::ops::Range;

//

macro_rules! some {
    ($v:expr) => {
        if let Some(v) = $v {
            return Some(v);
        }
    };
}

macro_rules! exhaustive {
    ($($tok:ident::$var:ident $({ $($t:tt)* })?,)*) => {{
        use crate::Token::*;

        #[allow(unreachable_code)]
        fn _assert_exhaustive() {
            match unreachable!() {
                $($var { .. } => {})*
            }
        }

        [
            $($tok::$var $({ $($t)* })?),*
        ]
    }};
}

//

pub type Result<T> = core::result::Result<T, Error>;

#[cfg_attr(test, derive(Serialize))]
#[derive(Debug, Clone, Copy)]
pub enum Error {
    UnexpectedEoi,
    ExtraTokens,
}

//

#[cfg_attr(test, derive(Serialize))]
#[derive(Debug, Clone, Copy)]
pub enum Token {
    /// `struct`
    Struct,

    /// `test`
    Test,

    /// identifiers like `_fn_name`
    Ident,

    /// `;`
    Semi,

    /// `:=`
    Init,

    /// `=`
    Assign,

    /// `(`
    LParen,

    /// `)`
    RParen,

    /// `{`
    Lbrace,

    /// `}`
    Rbrace,

    /// `+`
    Plus,

    /// `-`
    Minus,

    /// `*`
    Asterisk,

    /// `/`
    Slash,

    /// `->`
    Arrow,

    /// `:`
    Type,

    /// `@`
    Extern,

    /// `//`
    LineComment,

    /// integer literal like `0xFFu8`
    LitInt(i128),

    /// string literal like `"test"`
    LitStr,

    /// end of input
    Eoi,
}

impl Token {
    pub const fn enumerate() -> &'static [Self] {
        &exhaustive![
            Token::Struct,
            Token::Test,
            Token::Ident,
            Token::Semi,
            Token::Init,
            Token::Assign,
            Token::LParen,
            Token::RParen,
            Token::Lbrace,
            Token::Rbrace,
            Token::Plus,
            Token::Minus,
            Token::Asterisk,
            Token::Slash,
            Token::Arrow,
            Token::Type,
            Token::Extern,
            Token::LineComment,
            Token::LitInt { 0: 0 },
            Token::LitStr,
            Token::Eoi,
        ]
    }

    pub const fn as_token_str(self) -> Option<&'static str> {
        Some(match self {
            Token::Struct {} => "struct",
            Token::Test => "test",
            Token::Ident => return None,
            Token::Semi => ";",
            Token::Init => ":=",
            Token::Assign => "=",
            Token::LParen => "(",
            Token::RParen => ")",
            Token::Lbrace => "{",
            Token::Rbrace => "}",
            Token::Plus => "+",
            Token::Minus => "-",
            Token::Asterisk => "*",
            Token::Slash => "/",
            Token::Arrow => "->",
            Token::Type => ":",
            Token::Extern => "@",
            Token::LineComment { .. } => "//",
            Token::LitInt { .. } => return None,
            Token::LitStr => return None,
            Token::Eoi => "",
        })
    }
}

//

#[cfg_attr(test, derive(Serialize))]
#[derive(Debug, Clone)]
pub struct SpannedToken<'a> {
    token: Token,
    span: Span<'a>,
}

impl<'a> SpannedToken<'a> {
    pub fn token(&self) -> Token {
        self.token
    }

    pub fn span(&self) -> Range<usize> {
        self.span.span()
    }

    pub fn str(&self) -> &'a str {
        self.span.str()
    }
}

//

#[derive(Debug, Clone)]
pub struct Span<'a> {
    range: Range<usize>,
    source: &'a str,
}

impl<'a> Span<'a> {
    pub fn span(&self) -> Range<usize> {
        self.range.clone()
    }

    pub fn str(&self) -> &'a str {
        &self.source[self.span()]
    }
}

#[cfg(test)]
impl Serialize for Span<'_> {
    fn serialize<S>(&self, serializer: S) -> core::result::Result<S::Ok, S::Error>
    where
        S: serde::Serializer,
    {
        let mut s = serializer.serialize_struct("SpannedToken", 3)?;
        s.serialize_field("range", &format!("{:?}", self.span()))?;
        s.serialize_field("str", &self.str())?;
        s.end()
    }
}

//

#[derive(Debug, Clone)]
pub struct Lexer<'a> {
    source: &'a str,
    at: &'a str,
    eoi: Option<Token>,
    err: Result<()>,
}

impl<'a> Lexer<'a> {
    pub const fn new(source: &'a str) -> Self {
        Self {
            source,
            at: source,
            eoi: Some(Token::Eoi),
            err: Ok(()),
        }
    }

    pub const fn error(&self) -> Result<()> {
        self.err
    }

    fn src_index(&self) -> usize {
        self.at.as_ptr() as usize - self.source.as_ptr() as usize
    }

    fn advance_by(&mut self, len: usize) {
        self.at = &self.at[len..];
    }

    fn advance(&mut self, span: &Span) {
        self.advance_by(span.range.len())
    }

    fn span_to(&self, len: usize) -> Span<'a> {
        let at = self.src_index();
        Span {
            range: at..at + len,
            source: self.source,
        }
    }

    fn spanned_token_to(&mut self, token: Token, len: usize) -> SpannedToken<'a> {
        let span = self.span_to(len);
        self.advance(&span);
        SpannedToken { token, span }
    }

    fn try_match(&mut self, token: Token) -> Option<SpannedToken<'a>> {
        let str = token.as_token_str()?;
        if !self.at.starts_with(str) {
            return None;
        }

        Some(self.spanned_token_to(token, str.len()))
    }

    fn try_match_number(&mut self) -> Option<SpannedToken<'a>> {
        let first = self.at.chars().next()?;

        if !first.is_ascii_digit() {
            return None;
        }

        let mut dot_found = false;
        let term = self.at[1..]
            .find(|c: char| {
                !c.is_ascii_digit()
                    || if dot_found {
                        false
                    } else {
                        dot_found = true;
                        c == '.'
                    }
            })
            .map(|term| term + 1)
            .unwrap_or(self.at.len());

        let token = if dot_found {
            todo!("LitFloat")
        } else {
            let num = self.at[..term].parse::<i128>().unwrap(); // the unwrap should never fail
            Token::LitInt(num)
        };

        Some(self.spanned_token_to(token, term))
    }

    fn try_match_str(&mut self) -> Option<SpannedToken<'a>> {
        if !self.at.starts_with('"') {
            return None;
        }

        let Some(term) = self.at[1..].find('"') else {
            self.err = Err(Error::UnexpectedEoi);
            return self.eoi();
        };

        Some(self.spanned_token_to(Token::LitStr, term + 2))
    }

    fn try_match_ident(&mut self) -> Option<SpannedToken<'a>> {
        let first = self.at.chars().next()?;

        if !(first == '_' || first.is_alphabetic()) {
            return None;
        }

        let term = self.at[1..]
            .find(|c: char| !c.is_alphanumeric())
            .map(|term| term + 1)
            .unwrap_or(self.at.len());

        Some(self.spanned_token_to(Token::Ident, term))
    }

    fn try_match_comment(&mut self) -> Option<SpannedToken<'a>> {
        if !self.at.starts_with("//") {
            return None;
        }

        let term = self.at[2..]
            .lines()
            .next()
            .map(|line| line.len() + 2)
            .unwrap_or(self.at.len());

        Some(self.spanned_token_to(Token::LineComment, term))
    }

    fn eoi(&mut self) -> Option<SpannedToken<'a>> {
        let eoi = self.eoi.take()?;
        Some(self.spanned_token_to(eoi, 0))
    }
}

impl<'a> Iterator for Lexer<'a> {
    type Item = SpannedToken<'a>;

    fn next(&mut self) -> Option<Self::Item> {
        self.eoi.as_ref()?;
        self.at = self.at.trim();

        // keywords first, so that keywords have a priority over idents
        some!(self.try_match(Token::Struct));
        some!(self.try_match(Token::Test));

        // custom tokens
        some!(self.try_match_str());
        some!(self.try_match_ident());
        some!(self.try_match_comment());
        some!(self.try_match_number());

        // simple tokens in the order of more specific -> less specific
        // length of 2
        some!(self.try_match(Token::Init));
        some!(self.try_match(Token::Arrow));
        // length of 1
        some!(self.try_match(Token::Semi));
        some!(self.try_match(Token::Assign));
        some!(self.try_match(Token::LParen));
        some!(self.try_match(Token::RParen));
        some!(self.try_match(Token::Lbrace));
        some!(self.try_match(Token::Rbrace));
        some!(self.try_match(Token::Type));
        some!(self.try_match(Token::Extern));
        some!(self.try_match(Token::Plus));
        some!(self.try_match(Token::Minus));
        some!(self.try_match(Token::Asterisk));
        some!(self.try_match(Token::Slash));

        if !self.at.is_empty() {
            self.err = Err(Error::ExtraTokens);
        }

        self.eoi()
    }
}

//

#[cfg(test)]
mod tests {
    use crate::{Lexer, Result, SpannedToken, Token};

    use insta::assert_yaml_snapshot;

    fn lex(str: &str) -> (&str, Result<()>, Vec<SpannedToken>) {
        let mut lexer = Lexer::new(str);
        let tokens: Vec<_> = lexer.by_ref().collect();
        (str, lexer.error(), tokens)
    }

    #[test]
    fn lex_test() {
        assert_yaml_snapshot!(lex(
            "in->\"vali\"@->:d synt(ax} {but) t;h:e l=e:=xer should handle it"
        ));

        assert_yaml_snapshot!(lex("_a2"));
        assert_yaml_snapshot!(lex(""));
        assert_yaml_snapshot!(lex("\""));
        assert_yaml_snapshot!(lex("\"a"));
        assert_yaml_snapshot!(lex("\"a\""));

        let all_tokens: String = Token::enumerate()
            .iter()
            .filter_map(|tok| tok.as_token_str())
            .collect();

        assert_yaml_snapshot!(lex(&all_tokens));

        assert_yaml_snapshot!(lex("// comment\ncode"));

        assert_yaml_snapshot!(lex(include_str!("../../../tests/trivial")));
    }
}
