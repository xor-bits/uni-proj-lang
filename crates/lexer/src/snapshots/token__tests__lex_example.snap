---
source: crates/token/src/lib.rs
expression: "lex(include_str!(\"../../../tests/trivial\"))"
---
- "// immutable compile-time const\nval := { 4 };\n// or just:\n_val := 4;\n\n// function 'assignment'\nmain := () -> i32 {\n    print(\"Hello, world!\" + val);\n\treturn 0;\n};\n\n// automatic `return` without `;`\nmul_2 := (v: i32) -> i32 { v * 2 };\n\n// anonymous structs:\n_swap1 := (a: i32, b: i32) -> struct { a: i32, b: i32 } {\n\tstruct {\n\t\ta: b,\n\t\tb: a,\n\t}\n};\n// or tuples:\n_swap2 := (a: i32, b: i32) -> (i32, i32) {\n\t(b, a)\n};\n\n// references\n_swap3 := (a: &i32, b: &i32) {\n\tb, a = a, b;\n};\n\nextern := (i32) -> i32 @ __lang_test_extern;\n\n// generics, instanced lazily on first use\nsum := <T> (a: T, b: T) -> T {\n\ta + b\n};\n\n// unit tests\ntest \"unit test\" {\n\tassert(main() == 0);\n\tassert(mul_2(val) == 8);\n\n\ta, b := 4, 5;\n\t_swap3(&a, &b);\n\tassert((a, b) = (5, 4));\n\n\textern(42);\n\n\tassert(sum(\"tests\", \" worked\") == \"tests worked\");\n}\n"
- Ok:
    - token: LineComment
      span:
        range: 0..31
        str: // immutable compile-time const
    - token: Ident
      span:
        range: 32..35
        str: val
    - token: Init
      span:
        range: 36..38
        str: ":="
    - token: Lbrace
      span:
        range: 39..40
        str: "{"
    - token:
        LitInt: 4
      span:
        range: 41..42
        str: "4"
    - token: Rbrace
      span:
        range: 43..44
        str: "}"
    - token: Semi
      span:
        range: 44..45
        str: ;
    - token: LineComment
      span:
        range: 46..57
        str: "// or just:"
    - token: Ident
      span:
        range: 58..62
        str: _val
    - token: Init
      span:
        range: 63..65
        str: ":="
    - token:
        LitInt: 4
      span:
        range: 66..67
        str: "4"
    - token: Semi
      span:
        range: 67..68
        str: ;
    - token: LineComment
      span:
        range: 70..94
        str: "// function 'assignment'"
    - token: Ident
      span:
        range: 95..99
        str: main
    - token: Init
      span:
        range: 100..102
        str: ":="
    - token: LParen
      span:
        range: 103..104
        str: (
    - token: RParen
      span:
        range: 104..105
        str: )
    - token: Arrow
      span:
        range: 106..108
        str: "->"
    - token: Ident
      span:
        range: 109..112
        str: i32
    - token: Lbrace
      span:
        range: 113..114
        str: "{"
    - token: Ident
      span:
        range: 119..124
        str: print
    - token: LParen
      span:
        range: 124..125
        str: (
    - token: LitStr
      span:
        range: 125..140
        str: "\"Hello, world!\""
    - token: Plus
      span:
        range: 141..142
        str: +
    - token: Ident
      span:
        range: 143..146
        str: val
    - token: RParen
      span:
        range: 146..147
        str: )
    - token: Semi
      span:
        range: 147..148
        str: ;
    - token: Ident
      span:
        range: 150..156
        str: return
    - token:
        LitInt: 0
      span:
        range: 157..158
        str: "0"
    - token: Semi
      span:
        range: 158..159
        str: ;
    - token: Rbrace
      span:
        range: 160..161
        str: "}"
    - token: Semi
      span:
        range: 161..162
        str: ;
    - token: LineComment
      span:
        range: 164..197
        str: "// automatic `return` without `;`"
    - token: Ident
      span:
        range: 198..201
        str: mul
    - token: Ident
      span:
        range: 201..203
        str: _2
    - token: Init
      span:
        range: 204..206
        str: ":="
    - token: LParen
      span:
        range: 207..208
        str: (
    - token: Ident
      span:
        range: 208..209
        str: v
    - token: Type
      span:
        range: 209..210
        str: ":"
    - token: Ident
      span:
        range: 211..214
        str: i32
    - token: RParen
      span:
        range: 214..215
        str: )
    - token: Arrow
      span:
        range: 216..218
        str: "->"
    - token: Ident
      span:
        range: 219..222
        str: i32
    - token: Lbrace
      span:
        range: 223..224
        str: "{"
    - token: Ident
      span:
        range: 225..226
        str: v
    - token: Asterisk
      span:
        range: 227..228
        str: "*"
    - token:
        LitInt: 2
      span:
        range: 229..230
        str: "2"
    - token: Rbrace
      span:
        range: 231..232
        str: "}"
    - token: Semi
      span:
        range: 232..233
        str: ;
    - token: LineComment
      span:
        range: 235..256
        str: "// anonymous structs:"
    - token: Ident
      span:
        range: 257..263
        str: _swap1
    - token: Init
      span:
        range: 264..266
        str: ":="
    - token: LParen
      span:
        range: 267..268
        str: (
    - token: Ident
      span:
        range: 268..269
        str: a
    - token: Type
      span:
        range: 269..270
        str: ":"
    - token: Ident
      span:
        range: 271..274
        str: i32
    - token: Comma
      span:
        range: 274..275
        str: ","
    - token: Ident
      span:
        range: 276..277
        str: b
    - token: Type
      span:
        range: 277..278
        str: ":"
    - token: Ident
      span:
        range: 279..282
        str: i32
    - token: RParen
      span:
        range: 282..283
        str: )
    - token: Arrow
      span:
        range: 284..286
        str: "->"
    - token: Struct
      span:
        range: 287..293
        str: struct
    - token: Lbrace
      span:
        range: 294..295
        str: "{"
    - token: Ident
      span:
        range: 296..297
        str: a
    - token: Type
      span:
        range: 297..298
        str: ":"
    - token: Ident
      span:
        range: 299..302
        str: i32
    - token: Comma
      span:
        range: 302..303
        str: ","
    - token: Ident
      span:
        range: 304..305
        str: b
    - token: Type
      span:
        range: 305..306
        str: ":"
    - token: Ident
      span:
        range: 307..310
        str: i32
    - token: Rbrace
      span:
        range: 311..312
        str: "}"
    - token: Lbrace
      span:
        range: 313..314
        str: "{"
    - token: Struct
      span:
        range: 316..322
        str: struct
    - token: Lbrace
      span:
        range: 323..324
        str: "{"
    - token: Ident
      span:
        range: 327..328
        str: a
    - token: Type
      span:
        range: 328..329
        str: ":"
    - token: Ident
      span:
        range: 330..331
        str: b
    - token: Comma
      span:
        range: 331..332
        str: ","
    - token: Ident
      span:
        range: 335..336
        str: b
    - token: Type
      span:
        range: 336..337
        str: ":"
    - token: Ident
      span:
        range: 338..339
        str: a
    - token: Comma
      span:
        range: 339..340
        str: ","
    - token: Rbrace
      span:
        range: 342..343
        str: "}"
    - token: Rbrace
      span:
        range: 344..345
        str: "}"
    - token: Semi
      span:
        range: 345..346
        str: ;
    - token: LineComment
      span:
        range: 347..360
        str: "// or tuples:"
    - token: Ident
      span:
        range: 361..367
        str: _swap2
    - token: Init
      span:
        range: 368..370
        str: ":="
    - token: LParen
      span:
        range: 371..372
        str: (
    - token: Ident
      span:
        range: 372..373
        str: a
    - token: Type
      span:
        range: 373..374
        str: ":"
    - token: Ident
      span:
        range: 375..378
        str: i32
    - token: Comma
      span:
        range: 378..379
        str: ","
    - token: Ident
      span:
        range: 380..381
        str: b
    - token: Type
      span:
        range: 381..382
        str: ":"
    - token: Ident
      span:
        range: 383..386
        str: i32
    - token: RParen
      span:
        range: 386..387
        str: )
    - token: Arrow
      span:
        range: 388..390
        str: "->"
    - token: LParen
      span:
        range: 391..392
        str: (
    - token: Ident
      span:
        range: 392..395
        str: i32
    - token: Comma
      span:
        range: 395..396
        str: ","
    - token: Ident
      span:
        range: 397..400
        str: i32
    - token: RParen
      span:
        range: 400..401
        str: )
    - token: Lbrace
      span:
        range: 402..403
        str: "{"
    - token: LParen
      span:
        range: 405..406
        str: (
    - token: Ident
      span:
        range: 406..407
        str: b
    - token: Comma
      span:
        range: 407..408
        str: ","
    - token: Ident
      span:
        range: 409..410
        str: a
    - token: RParen
      span:
        range: 410..411
        str: )
    - token: Rbrace
      span:
        range: 412..413
        str: "}"
    - token: Semi
      span:
        range: 413..414
        str: ;
    - token: LineComment
      span:
        range: 416..429
        str: // references
    - token: Ident
      span:
        range: 430..436
        str: _swap3
    - token: Init
      span:
        range: 437..439
        str: ":="
    - token: LParen
      span:
        range: 440..441
        str: (
    - token: Ident
      span:
        range: 441..442
        str: a
    - token: Type
      span:
        range: 442..443
        str: ":"
    - token: Ampersand
      span:
        range: 444..445
        str: "&"
    - token: Ident
      span:
        range: 445..448
        str: i32
    - token: Comma
      span:
        range: 448..449
        str: ","
    - token: Ident
      span:
        range: 450..451
        str: b
    - token: Type
      span:
        range: 451..452
        str: ":"
    - token: Ampersand
      span:
        range: 453..454
        str: "&"
    - token: Ident
      span:
        range: 454..457
        str: i32
    - token: RParen
      span:
        range: 457..458
        str: )
    - token: Lbrace
      span:
        range: 459..460
        str: "{"
    - token: Ident
      span:
        range: 462..463
        str: b
    - token: Comma
      span:
        range: 463..464
        str: ","
    - token: Ident
      span:
        range: 465..466
        str: a
    - token: Assign
      span:
        range: 467..468
        str: "="
    - token: Ident
      span:
        range: 469..470
        str: a
    - token: Comma
      span:
        range: 470..471
        str: ","
    - token: Ident
      span:
        range: 472..473
        str: b
    - token: Semi
      span:
        range: 473..474
        str: ;
    - token: Rbrace
      span:
        range: 475..476
        str: "}"
    - token: Semi
      span:
        range: 476..477
        str: ;
    - token: Ident
      span:
        range: 479..485
        str: extern
    - token: Init
      span:
        range: 486..488
        str: ":="
    - token: LParen
      span:
        range: 489..490
        str: (
    - token: Ident
      span:
        range: 490..493
        str: i32
    - token: RParen
      span:
        range: 493..494
        str: )
    - token: Arrow
      span:
        range: 495..497
        str: "->"
    - token: Ident
      span:
        range: 498..501
        str: i32
    - token: Extern
      span:
        range: 502..503
        str: "@"
    - token: Ident
      span:
        range: 504..505
        str: _
    - token: Ident
      span:
        range: 505..510
        str: _lang
    - token: Ident
      span:
        range: 510..515
        str: _test
    - token: Ident
      span:
        range: 515..522
        str: _extern
    - token: Semi
      span:
        range: 522..523
        str: ;
    - token: LineComment
      span:
        range: 525..567
        str: "// generics, instanced lazily on first use"
    - token: Ident
      span:
        range: 568..571
        str: sum
    - token: Init
      span:
        range: 572..574
        str: ":="
    - token: Lt
      span:
        range: 575..576
        str: "<"
    - token: Ident
      span:
        range: 576..577
        str: T
    - token: Gt
      span:
        range: 577..578
        str: ">"
    - token: LParen
      span:
        range: 579..580
        str: (
    - token: Ident
      span:
        range: 580..581
        str: a
    - token: Type
      span:
        range: 581..582
        str: ":"
    - token: Ident
      span:
        range: 583..584
        str: T
    - token: Comma
      span:
        range: 584..585
        str: ","
    - token: Ident
      span:
        range: 586..587
        str: b
    - token: Type
      span:
        range: 587..588
        str: ":"
    - token: Ident
      span:
        range: 589..590
        str: T
    - token: RParen
      span:
        range: 590..591
        str: )
    - token: Arrow
      span:
        range: 592..594
        str: "->"
    - token: Ident
      span:
        range: 595..596
        str: T
    - token: Lbrace
      span:
        range: 597..598
        str: "{"
    - token: Ident
      span:
        range: 600..601
        str: a
    - token: Plus
      span:
        range: 602..603
        str: +
    - token: Ident
      span:
        range: 604..605
        str: b
    - token: Rbrace
      span:
        range: 606..607
        str: "}"
    - token: Semi
      span:
        range: 607..608
        str: ;
    - token: LineComment
      span:
        range: 610..623
        str: // unit tests
    - token: Test
      span:
        range: 624..628
        str: test
    - token: LitStr
      span:
        range: 629..640
        str: "\"unit test\""
    - token: Lbrace
      span:
        range: 641..642
        str: "{"
    - token: Ident
      span:
        range: 644..650
        str: assert
    - token: LParen
      span:
        range: 650..651
        str: (
    - token: Ident
      span:
        range: 651..655
        str: main
    - token: LParen
      span:
        range: 655..656
        str: (
    - token: RParen
      span:
        range: 656..657
        str: )
    - token: Assign
      span:
        range: 658..659
        str: "="
    - token: Assign
      span:
        range: 659..660
        str: "="
    - token:
        LitInt: 0
      span:
        range: 661..662
        str: "0"
    - token: RParen
      span:
        range: 662..663
        str: )
    - token: Semi
      span:
        range: 663..664
        str: ;
    - token: Ident
      span:
        range: 666..672
        str: assert
    - token: LParen
      span:
        range: 672..673
        str: (
    - token: Ident
      span:
        range: 673..676
        str: mul
    - token: Ident
      span:
        range: 676..678
        str: _2
    - token: LParen
      span:
        range: 678..679
        str: (
    - token: Ident
      span:
        range: 679..682
        str: val
    - token: RParen
      span:
        range: 682..683
        str: )
    - token: Assign
      span:
        range: 684..685
        str: "="
    - token: Assign
      span:
        range: 685..686
        str: "="
    - token:
        LitInt: 8
      span:
        range: 687..688
        str: "8"
    - token: RParen
      span:
        range: 688..689
        str: )
    - token: Semi
      span:
        range: 689..690
        str: ;
    - token: Ident
      span:
        range: 693..694
        str: a
    - token: Comma
      span:
        range: 694..695
        str: ","
    - token: Ident
      span:
        range: 696..697
        str: b
    - token: Init
      span:
        range: 698..700
        str: ":="
    - token:
        LitInt: 4
      span:
        range: 701..702
        str: "4"
    - token: Comma
      span:
        range: 702..703
        str: ","
    - token:
        LitInt: 5
      span:
        range: 704..705
        str: "5"
    - token: Semi
      span:
        range: 705..706
        str: ;
    - token: Ident
      span:
        range: 708..714
        str: _swap3
    - token: LParen
      span:
        range: 714..715
        str: (
    - token: Ampersand
      span:
        range: 715..716
        str: "&"
    - token: Ident
      span:
        range: 716..717
        str: a
    - token: Comma
      span:
        range: 717..718
        str: ","
    - token: Ampersand
      span:
        range: 719..720
        str: "&"
    - token: Ident
      span:
        range: 720..721
        str: b
    - token: RParen
      span:
        range: 721..722
        str: )
    - token: Semi
      span:
        range: 722..723
        str: ;
    - token: Ident
      span:
        range: 725..731
        str: assert
    - token: LParen
      span:
        range: 731..732
        str: (
    - token: LParen
      span:
        range: 732..733
        str: (
    - token: Ident
      span:
        range: 733..734
        str: a
    - token: Comma
      span:
        range: 734..735
        str: ","
    - token: Ident
      span:
        range: 736..737
        str: b
    - token: RParen
      span:
        range: 737..738
        str: )
    - token: Assign
      span:
        range: 739..740
        str: "="
    - token: LParen
      span:
        range: 741..742
        str: (
    - token:
        LitInt: 5
      span:
        range: 742..743
        str: "5"
    - token: Comma
      span:
        range: 743..744
        str: ","
    - token:
        LitInt: 4
      span:
        range: 745..746
        str: "4"
    - token: RParen
      span:
        range: 746..747
        str: )
    - token: RParen
      span:
        range: 747..748
        str: )
    - token: Semi
      span:
        range: 748..749
        str: ;
    - token: Ident
      span:
        range: 752..758
        str: extern
    - token: LParen
      span:
        range: 758..759
        str: (
    - token:
        LitInt: 42
      span:
        range: 759..761
        str: "42"
    - token: RParen
      span:
        range: 761..762
        str: )
    - token: Semi
      span:
        range: 762..763
        str: ;
    - token: Ident
      span:
        range: 766..772
        str: assert
    - token: LParen
      span:
        range: 772..773
        str: (
    - token: Ident
      span:
        range: 773..776
        str: sum
    - token: LParen
      span:
        range: 776..777
        str: (
    - token: LitStr
      span:
        range: 777..784
        str: "\"tests\""
    - token: Comma
      span:
        range: 784..785
        str: ","
    - token: LitStr
      span:
        range: 786..795
        str: "\" worked\""
    - token: RParen
      span:
        range: 795..796
        str: )
    - token: Assign
      span:
        range: 797..798
        str: "="
    - token: Assign
      span:
        range: 798..799
        str: "="
    - token: LitStr
      span:
        range: 800..814
        str: "\"tests worked\""
    - token: RParen
      span:
        range: 814..815
        str: )
    - token: Semi
      span:
        range: 815..816
        str: ;
    - token: Rbrace
      span:
        range: 817..818
        str: "}"
    - token: Eoi
      span:
        range: 818..818
        str: ""

